Source/FieldSolver/FiniteDifferenceSolver/FiniteDifferenceAlgorithms/CKCAlgorithm.H

```diff
@@ -0,0 +1,229 @@
+#        if defined WARPX_DIM_3D
+            const Real delta = std::max( { inv_dx,inv_dy,inv_dz } );
+            const Real rx = (inv_dx/delta)*(inv_dx/delta);
+            const Real ry = (inv_dy/delta)*(inv_dy/delta);
+            const Real rz = (inv_dz/delta)*(inv_dz/delta);
+            const Real beta = 0.125*(1. - rx*ry*rz/(ry*rz + rz*rx + rx*ry));
+            betaxy = ry*beta;
+            betaxz = rz*beta;
+            betayx = rx*beta;
+            betayz = rz*beta;
+            betazx = rx*beta;
+            betazy = ry*beta;
+            gammax = ry*rz*(0.0625 - 0.125*ry*rz/(ry*rz + rz*rx + rx*ry));
+            gammay = rx*rz*(0.0625 - 0.125*rx*rz/(ry*rz + rz*rx + rx*ry));
+            gammaz = rx*ry*(0.0625 - 0.125*rx*ry/(ry*rz + rz*rx + rx*ry));
+            alphax = 1. - 2.*betaxy - 2.*betaxz - 4.*gammax;
+            alphay = 1. - 2.*betayx - 2.*betayz - 4.*gammay;
+            alphaz = 1. - 2.*betazx - 2.*betazy - 4.*gammaz;
+            betaxy *= inv_dx;
+            betaxz *= inv_dx;
+            betayx *= inv_dy;
+            betayz *= inv_dy;
+            betazx *= inv_dz;
+            betazy *= inv_dz;
+            alphax *= inv_dx;
+            alphay *= inv_dy;
+            alphaz *= inv_dz;
+            gammax *= inv_dx;
+            gammay *= inv_dy;
+            gammaz *= inv_dz;
+#        elif defined WARPX_DIM_XZ
+            const Real delta = std::max(inv_dx,inv_dz);
+            const Real rx = (inv_dx/delta)*(inv_dx/delta);
+            const Real rz = (inv_dz/delta)*(inv_dz/delta);
+            betaxz = 0.125*rz;
+            betazx = 0.125*rx;
+            alphax = 1. - 2.*betaxz;
+            alphaz = 1. - 2.*betazx;
+            betaxz *= inv_dx;
+            betazx *= inv_dz;
+            alphax *= inv_dx;
+            alphaz *= inv_dz;
+#        endif
+
+        // Store the coefficients in array `stencil_coefs`, in prescribed order
+        stencil_coefs_x.resize(6);
+        stencil_coefs_x[0] = inv_dx;
+        stencil_coefs_x[1] = alphax;
+        stencil_coefs_x[2] = betaxy;
+        stencil_coefs_x[3] = betaxz;
+        stencil_coefs_x[4] = gammax;
+        stencil_coefs_y.resize(6);
+        stencil_coefs_y[0] = inv_dy;
+        stencil_coefs_y[1] = alphay;
+        stencil_coefs_y[2] = betayz;
+        stencil_coefs_y[3] = betayx;
+        stencil_coefs_y[4] = gammay;
+        stencil_coefs_z.resize(6);
+        stencil_coefs_z[0] = inv_dz;
+        stencil_coefs_z[1] = alphaz;
+        stencil_coefs_z[2] = betazx;
+        stencil_coefs_z[3] = betazy;
+        stencil_coefs_z[4] = gammaz;
+    }
+
+    /**
+    /* Perform derivative along x on a cell-centered grid, from a nodal field `F`*/
+    AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+    static amrex::Real UpwardDx(
+        amrex::Array4<amrex::Real> const& F,
+        amrex::Real const* coefs_x, int const n_coefs_x,
+        int const i, int const j, int const k ) {
+
+        amrex::Real alphax = coefs_x[1];
```

Comment:
important: `const`-ify those device variables:
```suggestion
        amrex::Real const alphax = coefs_x[1];
```